src/Microsoft.OData.Core/ODataBatchWriter.cs

```diff
@@ -802,55 +643,139 @@ private void VerifyCallAllowed(bool synchronousCall)
         }
 
         /// <summary>
-        /// Catch any exception thrown by the action passed in; in the exception case move the writer into
-        /// state ExceptionThrown and then rethrow the exception.
+        /// Verifies that calling CreateOperationRequestMessage is valid.
         /// </summary>
-        /// <param name="action">The action to execute.</param>
-        private void InterceptException(Action action)
+        /// <param name="synchronousCall">true if the call is to be synchronous; false otherwise.</param>
+        /// <param name="method">The Http method to be used for this request operation.</param>
+        /// <param name="uri">The Uri to be used for this request operation.</param>
+        /// <param name="contentId">The Content-ID value to write in ChangeSet head.</param>
+        private void VerifyCanCreateOperationRequestMessage(bool synchronousCall, string method, Uri uri,
+            string contentId)
         {
-            try
+            // Common verification.
+            this.ValidateWriterReady();
+            this.VerifyCallAllowed(synchronousCall);
+
+            if (this.outputContext.WritingResponse)
             {
-                action();
+                this.ThrowODataException(Strings.ODataBatchWriter_CannotCreateRequestOperationWhenWritingResponse);
             }
-            catch
+
+            ExceptionUtils.CheckArgumentStringNotNullOrEmpty(method, "method");
+
+            ExceptionUtils.CheckArgumentNotNull(uri, "uri");
+
+            // For the case within a changeset, verify CreateOperationRequestMessage is valid.
+            if (this.isInsideSubBatch)
             {
-                if (!IsErrorState(this.state))
+                if (HttpUtils.IsQueryMethod(method))
                 {
-                    this.SetState(BatchWriterState.Error);
+                    this.ThrowODataException(Strings.ODataBatch_InvalidHttpMethodForChangeSetRequest(method));
                 }

```

Comment:
Prohibiting requests within a changeset is also legacy, and should be under this.outputContext.MessageWriterSettings.Version <= ODataVersion.V4.

Reply:
Added AND clause for <=V4 for prohibition of queries within changeset.

---
In reply to: [145852599](https://github.com/OData/odata.net/pull/960#discussion_r145852599) [](ancestors = 145852599)